keps/sig-node/20190226-pod-overhead.md

```diff
@@ -0,0 +1,300 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Sandbox runtimes introduce a non-negligible overhead at the pod level which must be accounted for
+effective scheduling, resource quota management, and constraining.
+
+## Motivation
+
+Pods have some resource overhead. In our traditional linux container (Docker) approach,
+the accounted overhead is limited to the infra (pause) container, but also invokes some
+overhead accounted to various system components including: Kubelet (control loops), Docker,
+kernel (various resources), fluentd (logs). The current approach is to reserve a chunk
+of resources for the system components (system-reserved, kube-reserved, fluentd resource
+request padding), and ignore the (relatively small) overhead from the pause container, but
+this approach is heuristic at best and doesn't scale well.
+
+With sandbox pods, the pod overhead potentially becomes much larger, maybe O(100mb). For
+example, Kata containers must run a guest kernel, kata agent, init system, etc. Since this
+overhead is too big to ignore, we need a way to account for it, starting from quota enforcement
+and scheduling.
+
+### Goals
+
+* Provide a mechanism for accounting pod overheads which are specific to a given runtime solution
+
+### Non-Goals
+
+* Making runtimeClass selections
+
+## Proposal
+
+Augment the RuntimeClass custom resource definition and the `PodSpec` to introduce the field
+`Overhead *ResourceRequirements`. This field represents the overhead associated with running a pod
+for a given runtimeClass.  A mutating admission controller is introduced which will update the `Overhead`
+field in the workload's `PodSpec` to match what is provided for the selected RuntimeClass, if one is specified.
+
+Kubelet's creation of the pod cgroup will be calculated as the sum of container `ResourceRequirements` fields,
```

Comment:
do we need to account for overhead in how we handle eviction decisions?  right now, we evict based on usage relative to request.  this may prove problematic as you will potentially want to subtract overhead from the observed usage to make a more accurate decision.

Reply:
I made this more explicit in the proposal.